File: frontend/lib/insights-defaults.ts
// Adaptive default window/baseline selection for the Insights page.
//
// The backend skips an insight (renders an info placeholder "Requires Nh of
// historical data") whenever the chosen baseline period is longer than the
// data that actually exists. So a brand-new service with the static 7-day
// baseline default sees almost every card degrade to "not enough data". This
// module picks the best window+baseline pair for how much history a service
// actually has, using ONLY values that the existing dropdowns + backend accept
// (window_size_hrs ∈ {0.25,1,4,24}; baseline_hours ∈ {1,24,168,720}). The
// pure functions here are the single source of truth shared by the page, the
// useInsightsDefaults hook, and the unit tests.

export const WINDOW_OPTIONS = [
  { label: 'Last 15 Minutes', value: '0.25' },
  { label: 'Last 1 Hour', value: '1' },
  { label: 'Last 4 Hours', value: '4' },
  { label: 'Last 24 Hours', value: '24' },
]

export const BASELINE_OPTIONS = [
  { label: 'Previous 1 Hour', value: '1' },
  { label: 'Last 24 Hours', value: '24' },
  { label: 'Last 7 Days', value: '168' },
  { label: 'Last 30 Days', value: '720' },
]

// The historical default (also the backend default): recent 1h vs previous 7d.
// Used whenever we have no extents to adapt to, so behavior is unchanged for
// services with no data and on the server/first render.
export const STATIC_DEFAULT = { window: '1', baseline: '168' } as const

// Hours of history available = now − earliest_log_at. The baseline period the
// Insights query compares against reaches back (now − baseline − window), so
// "how far back data goes from now" is exactly the quantity a baseline must
// fit inside. Returns null when there's no/unparseable extent (→ STATIC_DEFAULT).
export function historyHoursFromExtents(
  earliest: string | null | undefined,
  now: number = Date.now(),
): number | null {
  if (!earliest) return null
  // Mirror FilterBar / useDataWindowOverlap: a date-only extent ("2026-06-15")
  // is widened to UTC start-of-day before parsing so a non-UTC machine doesn't
  // mis-bucket the span.
  const iso = earliest.length === 10 ? earliest + 'T00:00:00.000Z' : earliest
  const ms = Date.parse(iso)
  if (!Number.isFinite(ms)) return null
  return Math.max(0, (now - ms) / 3_600_000)
}

// Map available history (hours) to the best { window, baseline } the existing
// dropdown options allow. Half-open intervals: an exact boundary (e.g. 24h)
// selects the higher bucket. Returns STATIC_DEFAULT for null/NaN so no-data
// services are unchanged.
export function pickInsightsDefault(
  historyHours: number | null,
): { window: string; baseline: string } {
  if (historyHours == null || !Number.isFinite(historyHours)) {
    return { window: STATIC_DEFAULT.window, baseline: STATIC_DEFAULT.baseline }
  }
  const h = historyHours
  if (h < 1) return { window: '0.25', baseline: '1' } // <1h: smallest pair; cards light up once 1h elapses
  if (h < 4) return { window: '1', baseline: '1' } //  ~1–4h: this hour vs the previous hour
  if (h < 24) return { window: '4', baseline: '1' } //  ~4–24h: last 4h vs previous hour
  if (h < 48) return { window: '4', baseline: '24' } // ~1–2d: last 4h vs previous day
  if (h < 168) return { window: '24', baseline: '24' } // ~2–7d: day over day
  if (h < 720) return { window: '1', baseline: '168' } // ≥7d: today's static default (1h vs 7d)
  return { window: '1', baseline: '720' } //                   ≥30d: 1h vs 30d
}

File: backend/utils/insights_defaults.py
"""Backend mirror of the frontend's adaptive Insights default picker.

PARITY CONTRACT (load-bearing): ``frontend/lib/insights-defaults.ts``
(``historyHoursFromExtents`` + ``pickInsightsDefault``) decides which
window/baseline pair the Insights page requests by default, based on how
much log history the active service has. The ``/api/insights`` cache key
folds in both hour values (``backend/repositories/insights/repository.py``),
so the prewarmer must warm EXACTLY the pair the page will pick — warming any
other pair is a guaranteed cache miss for every default page load. That is
the regression this module fixes: once a service accrues ≥720 h (30 d) of
history the page defaults to (1 h, 720 h), while the prewarmer warmed only
the static (1 h, 168 h), so the default selection cold-computed (~20 s on a
long-history service) on every visit outside the cache TTL.

Both roles resolve the same bucket: the extents the client picks from
(bootstrap ``log_extents`` / ``/api/log-extents``) are the service-wide
snapshot, un-clamped for analysts, so one derivation per service covers the
admin warm and every analyst-shape warm.

Keep the band table here identical to the one in
``frontend/lib/insights-defaults.ts``. ``tests/utils/test_insights_defaults.py``
pins every band AND parses the TS source so the two sides cannot drift
silently.
"""

from __future__ import annotations

import math
from datetime import UTC, datetime

from backend.utils.date_utils import parse_iso_utc

# The historical static default (also the InsightsRequest field defaults):
# recent 1 h vs previous 7 d. Used when a service has no extents yet, so
# no-data services behave exactly as before.
STATIC_DEFAULT: tuple[float, float] = (1.0, 168.0)

# (history_upper_bound_hours, window_hours, baseline_hours). Half-open bands
# over available history — an exact boundary selects the higher bucket, and
# the final +inf band is the ≥30 d shape. Values are floats because the
# client sends ``parseFloat`` of the dropdown tokens, so these produce
# byte-identical cache-key fragments.
_BANDS: tuple[tuple[float, float, float], ...] = (
    (1.0, 0.25, 1.0),
    (4.0, 1.0, 1.0),
    (24.0, 4.0, 1.0),
    (48.0, 4.0, 24.0),
    (168.0, 24.0, 24.0),
    (720.0, 1.0, 168.0),
    (math.inf, 1.0, 720.0),
)


def history_hours_from_earliest(earliest: str | None, *, now: datetime | None = None) -> float | None:
    """Hours of history available = ``now − earliest_log_at``, floored at 0.

    Mirrors ``historyHoursFromExtents``: ``parse_iso_utc`` widens a date-only
    extent ("2026-06-15") to UTC start-of-day the same way the TS side does;
    absent/unparseable extents return None (→ ``STATIC_DEFAULT``).
    """
    dt = parse_iso_utc(earliest)
    if dt is None:
        return None
    anchor = now if now is not None else datetime.now(UTC)
    return max(0.0, (anchor - dt).total_seconds() / 3600.0)


def pick_insights_default(history_hours: float | None) -> tuple[float, float]:
    """Map available history to the default ``(window_hours, baseline_hours)``."""
    if history_hours is None or not math.isfinite(history_hours):
        return STATIC_DEFAULT
    for upper, window, baseline in _BANDS:
        if history_hours < upper:
            return (window, baseline)
    return STATIC_DEFAULT  # unreachable: the last band's bound is +inf
